### DISCUSSION TITLE
How to eliminate this threat for sqlalchemy

### DISCUSSION THREAD

**Discussion by zhangyanli-pro** [sqlalchemy/sqlalchemy#8852] [Usage Questions]
The sqlalchemy package is vulnerable to Deserialization of Untrusted Data leading to Arbitrary Code Execution. The loads() method in the serializer.py file will attempt to serialize custom objects with no preset serialization mechanism, resorting to utilizing their __reduce__() methods to perform the action. The unsafe deserialization occurs when user-provided external objects are serialized. A remote attacker may craft invalid source code and embed a payload inside a custom __reduce__() method to be run when the data is serialized by the sqlalchemy application. By enticing a victim to serialize the invalid code, the attacker can achieve code execution on systems that the victim controls. Attackers may also take advantage of this vulnerability if an external object written by a user allows an attacker to control input that is serialized by the sqlalchemy library.

**Comment by CaselIT** [comment]
Hi,

The serializer is an extension to sqlalchemy, so application need to explicitly opt into using it. If you are in a situation where the input cannot be trusted then that extension is probably not advisable.

On sqlalchemy site I think we can add a note in the documentation mentioning this security consideration, that is currently missing. @zzzeek

**Comment by zzzeek** [comment]
serializer is an extension to Python pickle.  The security issues with pickle are [well known and documented](https://docs.python.org/3/library/pickle.html).

**Reply by CaselIT** [reply]
Sure, I meant that we could add a note in the docs better indicate this, maybe with a link to the python docs

**Reply by CaselIT** [reply]
created https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4228
